ek-computation/src/controller/executor.rs

```
@@ -295,12 +295,23 @@ impl NaiveExecutor {
     }
 
     #[instrument(skip(self, req))]
-    fn break_down_to_egress(&mut self, req: &v1::ForwardReq, req_id: ReqId) {
+    async fn break_down_to_egress(&mut self, req: &v1::ForwardReq, req_id: ReqId) {
         for (idx, seq) in req.sequences.iter().enumerate() {
             // update pending_seq
             let seq_gid = self.add_seq(req_id, idx as LocalSeqIdx);
             // update pending_req
             for (idx, expert) in seq.experts.iter().enumerate() {
+                if self
+                    .registry
+                    .lock()
```

Acquire a lock here will introduce too much overhead. Expert existence check should not be placed here.

---

Besides, we should avoid using async function if possible.

---

Thanks for your comments. What about using a rwlock (to reduce overhead)? Which place to check existency you think is better?

---

My opinion is avoiding check here. If non-existent expert is used, it must be an implementation or configuration error which can not be recovered, so let it fail is better.

---

So I think we can simply let rust panic when we meet these nonexistent experts.

---

No need to panic here cause we can not assert the error is in the controller binary. Based on my experience, this error is more likely caused by the integration plugin.

---

The issue states when nonexistent experts detected,`pending_egress` always contains them (instead of removing them) and leads to incessive annoying errors in log. Maybe we can remove them [here](https://github.com/expert-kit/expert-kit/blob/647a8e56133d86cbf2c3aaa120ffb34797e2d76c/ek-computation/src/controller/executor.rs#L159), like:

```rust
let Ok(channel) = self.registry.lock().await.select(expert_id).await else {
    // remove the expert_id from pending_egress
    // and log::warn! or log::error!
    continue; // or return error to the caller?
};
```